File: Bookcase core/src/main/java/cz/vhromada/book/service/BookService.java
package cz.vhromada.book.service;

import java.util.ArrayList;

import cz.vhromada.book.domain.Book;
import cz.vhromada.book.repository.BookRepository;
import cz.vhromada.common.service.AbstractMovableService;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.Cache;
import org.springframework.stereotype.Component;

/**
 * A class represents implementation of service for books.
 *
 * @author Vladimir Hromada
 */
@Component("bookService")
public class BookService extends AbstractMovableService<Book> {

    /**
     * Creates a new instance of BookService.
     *
     * @param bookRepository repository for books
     * @param cache          cache
     * @throws IllegalArgumentException if repository for books is null
     *                                  or cache is null
     */
    @Autowired
    public BookService(final BookRepository bookRepository, @Value("#{cacheManager.getCache('bookcaseCache')}") final Cache cache) {
        super(bookRepository, cache, "books");
    }

    @Override
    protected Book getCopy(final Book data) {
        final Book newBook = new Book();
        newBook.setCzechName(data.getCzechName());
        newBook.setOriginalName(data.getOriginalName());
        newBook.setLanguages(new ArrayList<>(data.getLanguages()));
        newBook.setIsbn(data.getIsbn());
        newBook.setIssueYear(data.getIssueYear());
        newBook.setDescription(data.getDescription());
        newBook.setElectronic(data.getElectronic());
        newBook.setPaper(data.getPaper());
        newBook.setNote(data.getNote());
        newBook.setPosition(data.getPosition());
        newBook.setAuthors(new ArrayList<>(data.getAuthors()));
        newBook.setCategories(new ArrayList<>(data.getCategories()));

        return newBook;
    }

}

File: bookcase-core/src/main/kotlin/cz/vhromada/bookcase/service/BookService.kt
package cz.vhromada.bookcase.service

import cz.vhromada.bookcase.domain.Book
import cz.vhromada.bookcase.repository.BookRepository
import cz.vhromada.common.entity.Account
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.AbstractMovableService
import cz.vhromada.common.utils.sorted
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.Cache
import org.springframework.stereotype.Component

/**
 * A class represents of service for books.
 *
 * @author Vladimir Hromada
 */
@Component("bookService")
class BookService(
        private val bookRepository: BookRepository,
        accountProvider: AccountProvider,
        timeProvider: TimeProvider,
        @Value("#{cacheManager.getCache('bookcaseCache')}") cache: Cache) : AbstractMovableService<Book>(bookRepository, accountProvider, timeProvider, cache, "books") {

    override fun getAccountData(account: Account): List<Book> {
        return bookRepository.findByAuditCreatedUser(account.id)
    }

    override fun getCopy(data: Book): Book {
        return data.copy(id = null, authors = data.authors.map { it }, categories = data.categories.map { it }, items = data.items.map { it.copy(id = null) })
    }

    override fun updatePositions(data: List<Book>) {
        val audit = getAudit()
        for (i in data.indices) {
            val book = data[i]
            book.position = i
            book.modify(audit)
            val items = book.items.sorted()
            for (j in items.indices) {
                val item = items[j]
                item.position = j
                item.modify(audit)
            }
        }
    }

}
